### DISCUSSION TITLE
How to integrate prosemirror-math api in tiptap v2

### DISCUSSION THREAD

**Discussion by rajeshtva** [ueberdosis/tiptap#1975] [General Questions]
currently i need to integrate prosemirror-math in tiptap v2 beta. i looked into tiptap document for integrating schema and nodes. But i didn't get it enough to implement it there. 

so does anyone suggest how can i achieve it? have anyone anything written regarding it?

may be i will need to use tiptap v1 for this. any suggestions?

**Comment by cadars** [comment]
This project uses math-inline and tiptap, maybe you can learn something from their implementation? https://github.com/AdminScientificHub/community/tree/03f4b72ef77e1301f93c88d090d7af59656d03ac/src/components/_common/TextEditor/_extensions/MathInline

Also check this issue, someone almost made it work: https://github.com/benrbray/prosemirror-math/issues/27

**Comment by rajeshtva** [comment]
@cadars  currently i am working on modifying @BrianHung's [gist](https://gist.github.com/BrianHung/b72126c98fa08cb1c09170b1394771a0)

```javascript
import { InputRule } from "prosemirror-inputrules";
import { Node } from "@tiptap/core";

import katex from "katex";
import "katex/dist/katex.min.css";
import "~/assets/css/Math.css";
import { deleteMath } from "./MathKeymaps.js";

const CustomNode = Node.create({
  name: "math",
  //   draggable: true,
  code: true,
  content: "text*",
  marks: "",
  group: "inline",
  inline: true,
  defining: true,
  isolating: true,
  parseHTML: [{ tag: "span.Math" }],
  renderHTML: node => {
    [
      "span",
      { class: "Math" },
      ["span", { class: "katex-render", contenteditable: "false" }],
      ["span", { contenteditable: "false" }, "$"],
      ["span", { class: "katex-editor" }, 0],
      ["span", { contenteditable: "false" }, "$"]
    ];
  },

  addInputRules({ type, getAttrs }) {
    return [
      new InputRule(
        /(?:\$)([^\$\s]+(?:\s+[^\$\s]+)*)(?:\$)$/,
        (state, match, start, end) => {
          const attrs =
            getAttrs instanceof Function ? getAttrs(match) : getAttrs;
          const [matchedText, content] = match;
          const { tr, schema } = state;
          if (matchedText)
            // Create the new Math node.
            tr.replaceWith(
              start,
              end,
              type.create(attrs, schema.text(content))
            );
          return tr;
        }
      )
    ];
  },

  addNodeView() {
    return {
      name: "Math",
      props: ["node", "view", "getPos"],
      computed: {
        active() {
          return this.parentHasSelection() ? "active" : "hidden";
        },
        hidden() {
          return this.parentHasSelection() ? "hidden" : "active";
        }
      },
      watch: {
        "node.textContent": function(textContent) {
          this.render(textContent);
        }
      },
      mounted() {
        if (this.node && this.node.textContent)
          this.render(this.node.textContent);
      },
      methods: {
        // Updates katex-render with node textContent.
        render(textContent) {
          katex.render(textContent, this.$refs.render, {
            throwOnError: false,
            displayMode: false
          });
        },
        // Shows katex-render and hides katex-editor when selection is on parent.
        parentHasSelection() {
          const {
            doc,
            selection: { from, to, anchor }
          } = this.view.state;
          const rpos = doc.resolve(this.getPos());
          const parentNodeFrom = this.getPos() - rpos.parentOffset;
          const parentNodeTo = parentNodeFrom + rpos.parent.nodeSize;
          const hasAnchor = parentNodeFrom <= anchor && anchor < parentNodeTo;
          const hasSelect = from < parentNodeTo && parentNodeFrom <= to;
          return hasAnchor || hasSelect;
        }
      },
      template: `
        <span class="Math">
          <span class="katex-render" v-bind:class="hidden" ref="render" v-show="!parentHasSelection()" contenteditable="false"></span><span :contenteditable="false" class="decoration-inline-math" v-bind:class="active">$</span><span class="katex-editor" v-bind:class="active" ref="content"></span><span :contenteditable="false" class="decoration-inline-math" v-bind:class="active">$</span>
        </span>
      `
    };
  },
  addKeyboardShortcuts() {
    return {
      Backspace: deleteMath
    };
  }
});

export default CustomNode;
```

but i am getting issue in `addInputRules` key. it is not passed  `type` and `getAttrs` method. 

any suggestions ?

**Reply by cadars** [reply]
Are you using the latest tiptap version?
Input rules are now part of tiptap (see https://github.com/ueberdosis/tiptap/pull/1997), maybe it's related to the issue?

**Reply by rajeshtva** [reply]
updated to new version. but issues still persists. the content inside `addInputRules` is taken directly from @BrianHung 's  [gist](https://gist.github.com/BrianHung/b72126c98fa08cb1c09170b1394771a0) i mentioned along with code and that is written for tiptap v1. I am migrating actually.  and i am having hard time relating old apis to new apis. i guess in v1 `inputRules` functions is passed some arguments that is no more passed in v2. in the migration guide it is nowhere mentioned.

i request you to look at my implementaiton of input rule and suggest change there. please also have a look at @BrianHung  [gist](https://gist.github.com/BrianHung/b72126c98fa08cb1c09170b1394771a0)

**edit**: `textblockTypeInputRule` in the addInputRules method has `type`, `find` `getAttributes` method. somewhat they seem similar. yet i can't find any way to access underlying schema & state. so that i can match logic

**Reply by BrianHung** [reply]
@je-rajesh I would definitely try using `prosemirror-math` over my gist 😅

**Reply by rajeshtva** [reply]
i changed my username from @je-rajesh to @raekcya

**Comment by rajeshtva** [comment]
@BrianHung  would you show how can i integrate prosemirror-math with tiptap if you have done already

**Reply by BrianHung** [reply]
@raekcya

1. With `addProseMirrorPlugins`, add https://github.com/benrbray/prosemirror-math/blob/master/src/math-plugin.ts
2. From https://github.com/benrbray/prosemirror-math/blob/master/src/math-schema.ts, copy the schema spec and create nodes for them.
